File: Napakalaki1/src/napakalaki1/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki1;
import GUI.Dice;
import java.util.ArrayList;

/**
 *
 * @author TOSHIBA
 */
public class Player {
    static final int  MAXLEVEL= 10;
    private String name;
    private int level;
    private boolean dead=true;
    private boolean canISteal=true;
    private ArrayList<Treasure> hiddenTreasures=new ArrayList();
    private ArrayList<Treasure> visibleTreasures=new ArrayList();
    private BadConsequence pendingBadConsequence;
    public int cLSectario;
    protected Player enemy;
    public boolean esSectario=false;
    
    public Player(String n){
        name=n;
        level=1;
        dead=false;
}
    public Player (Player p){
        this.name = p.name;
        this.level = p.level;
        this.pendingBadConsequence = p.pendingBadConsequence;
        this.dead = p.dead;
        this.hiddenTreasures = p.hiddenTreasures;
        this.visibleTreasures = p.visibleTreasures;
        this.enemy = p.enemy;
        this.canISteal= p.canISteal;
        this.esSectario=p.esSectario;
       
    }
    public BadConsequence getPending(){
        return pendingBadConsequence;
    }
    protected Player getEnemy(){
        return enemy;
    }
    public String nombreEnemigo(){
        return enemy.getName();
    }
    public String getName(){
        return name;
    }
    private  void bringToLife(){
        if(dead==true)
            dead=false;
    }
    public int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    public boolean shouldConvert(){
        Dice d = Dice.getInstance();
        int num = d.nextNumber();
        if (num == 1) {
            return true;
        }
        else{
            return false;
        }
    }
    public int getCombatLevel(){
  
        int nivel=level;
        
        for(Treasure v : visibleTreasures){
        nivel=nivel+v.getBonus();  
        }
        return nivel;
        
     }
    
    public int GetCombatLevel(){
        int nivel=level;
        
        for(Treasure v : visibleTreasures){
        nivel=nivel+v.getBonus();  
        }
        return nivel;
    }
    private void incrementLevels(int i){
        level=level+i;
    }
    private void decrementLevels(int i){
        level=level-i;
        if(level<1)
            level=1;
    }
    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence=b;
    }
    private void applyPrize(Monster m){
        int nlevels = m.getLevelsGained();
        incrementLevels(nlevels);
        int ntreasure = m.getTreasuresGained();
        int num= hiddenTreasures.size();
        if(ntreasure>0){
            if(num+ntreasure<5){
            CardDealer dealer = CardDealer.getInstance();
               for(int i=0; i<ntreasure; i++){
                    Treasure treasure = dealer.nextTreasure();
                 hiddenTreasures.add(treasure);            
                 }
            }
            else{
            CardDealer dealer = CardDealer.getInstance();
                for(int i=num; i<4; i++){
                    Treasure treasure = dealer.nextTreasure();
                 hiddenTreasures.add(treasure);            
                 }
            }
        }
    }
    private void applyBadConsequence(Monster m){
        BadConsequence badConsequence = m.getBadConsequence();
         /*   for(Treasure t:visibleTreasures){
                discardVisibleTreasure(t);
            }
            for(Treasure tr:hiddenTreasures){
                discardHiddenTreasure(tr);
            }
        }*/
        int nlevels = badConsequence.getLevels();
        decrementLevels(nlevels);
        BadConsequence pendingBad = badConsequence.adjustToFitTreasureLists(visibleTreasures,hiddenTreasures);
        setPendingBadConsequence(pendingBad);
    }
    private boolean canMakeTreasureVisible(Treasure t){
        boolean puede=true;
        TreasureKind tipo=t.getType();
        if(tipo==TreasureKind.ONEHAND || tipo==TreasureKind.BOTHHANDS){
            int contador_hand=0;
            for(int i=0;i<visibleTreasures.size() && puede;i++){
                if(visibleTreasures.get(i).getType()==TreasureKind.BOTHHANDS){
                    puede=false;
                }
                else if(visibleTreasures.get(i).getType()==TreasureKind.ONEHAND){
                    contador_hand++;
                    if(tipo==TreasureKind.BOTHHANDS) puede=false;
                }
                if(contador_hand==2) puede=false;
            }
        }
        else{
            for(int i=0;i<visibleTreasures.size() && puede;i++){
                if(tipo==visibleTreasures.get(i).getType()){
                    puede=false;
                }
            }
        }
        return puede;
    }
    private int howManyVisibleTreasures(TreasureKind tKind){
        int j=0;
        for(int i=0;i<visibleTreasures.size();i++){
            if(visibleTreasures.get(i).getType()==tKind){
                j++;
            }
        }
        return j;
    }
    private void dieIfNoTreasures(){
        if(hiddenTreasures.isEmpty() && visibleTreasures.isEmpty())
            dead=true;
    }
    public boolean isDead(){
        return dead;
    }
    public ArrayList<Treasure> getHiddenTreasures(){
        
        return hiddenTreasures;
    }
    public ArrayList <Treasure> getVisibleTreasures(){
        return visibleTreasures;
    }
 
    public CombatResult combat(Monster m){
        CombatResult combatResult;
        int myLevel = getCombatLevel();
        int monsterLevel = getOponentLevel(m);
        Napakalaki napakalaki= Napakalaki.getInstance();
        if(myLevel >monsterLevel){
            applyPrize(m);
            if(getLevels()>=MAXLEVEL)
                combatResult = CombatResult.WINGAME;
            else
                combatResult = CombatResult.WIN;
        }
        else{          
            applyBadConsequence(m);
            combatResult = CombatResult.LOSE;
            if(shouldConvert()){
                System.out.println(name);
                combatResult=CombatResult.LOSEANDCONVERT;
                esSectario=true;
            }
        }
        return combatResult;
    }
    public void makeTreasureVisible (Treasure t){
        boolean canI=canMakeTreasureVisible(t);
        if(canI){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }     
    }
    public void discardVisibleTreasure (Treasure t){
        visibleTreasures.remove(t);
       if(pendingBadConsequence!=null && !pendingBadConsequence.isEmpty()){
           pendingBadConsequence.substractVisibleTreasure(t);
       }
       dieIfNoTreasures();
    }
    public void discardHiddenTreasure (Treasure t){
        hiddenTreasures.remove(t);
        if(pendingBadConsequence!=null && !pendingBadConsequence.isEmpty()){
            pendingBadConsequence.substractHiddenTreasure(t);
        }
        dieIfNoTreasures();
    }
    public boolean validState(){
        //if(/*pendingBadConsequence.isEmpty() &&*/ hiddenTreasures.size()<=4)
        if (pendingBadConsequence == null || (this.pendingBadConsequence.isEmpty() && this.hiddenTreasures.size() <= 4)) 
            return true;
        else
            return false;
    }
    public void initTreasures(){
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        bringToLife();
        Treasure treasure = dealer.nextTreasure();
        hiddenTreasures.add(treasure);
        int number = dice.nextNumber();
        
        if(number==6){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
        if(number>1 && number <6){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
    }
    public int getLevels(){
        return level;
    }
    public Treasure stealTreasure(){
        boolean canI = canISteal();
          Treasure treasure = null;
        if(canI){
            boolean canYou = canYouGiveMeATreasure();
            if(canYou){
                treasure = enemy.giveMeATreasure();
                hiddenTreasures.add(treasure);
                haveStolen();
            }
        }
        return treasure;
    }
    public void setEnemy(Player enemi){
        enemy=enemi;
    }
    private Treasure giveMeATreasure(){
        int random = (int) (Math.random()*hiddenTreasures.size());
        return hiddenTreasures.get(random);
        
    }
    public boolean canISteal(){ 
        if(enemy.hiddenTreasures.size()>0){
            return true;
        }
        else{
            return false;
        }
    }
    private boolean canYouGiveMeATreasure(){
        if(hiddenTreasures.isEmpty())
            return false;
        else
            return true;
    }
    private void haveStolen(){
        canISteal=false;
    }
    public void discardAllTreasures(){     
        for(Treasure t : visibleTreasures)
            discardVisibleTreasure(t);
        for(Treasure tr : hiddenTreasures)
            discardHiddenTreasure(tr);
    }
}

File: NapakalakiRuby1/lib/player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'card_dealer.rb'
require_relative 'bad_consequence.rb'
require_relative 'treasure.rb'
require_relative 'treasure_kind.rb'
require_relative 'dice.rb'

module NapakalakiGame
  class Player
    attr_reader :name
    attr_accessor :dead, :level, :visibleTreasures, :hiddenTreasures, :pendingBadConsequence, :enemy, :canISteal
    @MAXLEVEL = 10
    def initialize (n)
      @name = n
      @dead = true
      @level = 1
      @visibleTreasures = Array.new
      @hiddenTreasures = Array.new
      @pendingBadConsequence = NumericBadConsequence.new("", 0, 0, 0)
      @enemy=nil
      @canISteal=true
    end
    def getName
      return @name
    end
    def getVisibleTreasures
      return @visibleTreasures
    end
    def getHiddenTreasures
      return @hiddenTreasures
    end
    def copia(p)
      @name=p.name
      @level=p.level
      @dead=p.dead
      @visibleTreasures=p.visibleTreasures
      @hiddenTreasures=p.hiddenTreasures
      @enemy=p.enemy
      @canISteal=p.canISteal
      @pendingBadConsequence=p.pendingBadConsequence
    end
    def self.MAXLEVEL
      return @MAXLEVEL
    end
    def bringToLife
      @dead = false
    end
    def getCombatLevel
      suma = @level
      @visibleTreasures.each do |element|
        suma += element.bonus
      end
      return suma
    end
    def getOponentLevel (m)
      return m.combatLevel
    end    
    def canISteal
      return @canISteal
    end
    def shouldConvert
      return Dice.instance.nextNumber == 1
    end
    def incrementLevels (i)
      @level += i
    end
    def setEnemy(e)
      @enemy=e
    end
    def decrementLevels (i)
      if (@level - i) <= 1
        @level = 1
      else
        @level -= i
      end
    end
    def setPendingBadConsequence (b)
      @pendingBadConsequence = b
    end
    def applyPrize (m)
      nLevels = m.getLevelsGained
      incrementLevels(nLevels)
      nTreasures = m.getTreasuresGained

      if nTreasures > 0
        dealer = CardDealer.instance

        for i in 1..nTreasures do
          treasure = dealer.nextTreasure
          @hiddenTreasures << treasure
        end
      end
    end
    def applyBadConsequence (m)
      badConsequence = m.badConsequence
      nLevels = badConsequence.level
      decrementLevels(nLevels)
      pendingBad = badConsequence.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
      setPendingBadConsequence(pendingBad)
    end
    def canMakeTreasureVisible (t)
      puede=false
      if t.type == TreasureKind::ONEHAND
        if howManyVisibleTreasures(TreasureKind::ONEHAND)<2 && howManyVisibleTreasures(TreasureKind::BOTHHANDS)==0
          puede=true
        end
      end
      if t.type == TreasureKind::BOTHHANDS
        if howManyVisibleTreasures(TreasureKind::ONEHAND)==0 && howManyVisibleTreasures(TreasureKind::BOTHHANDS)==0
          puede=true
        end
      end
      if t.type == TreasureKind::HELMET
        if howManyVisibleTreasures(TreasureKind::HELMET)==0
          puede=true
        end
      end
      if t.type == TreasureKind::SHOES
        if howManyVisibleTreasures(TreasureKind::SHOES)==0
          puede=true
        end
      end
      if t.type == TreasureKind::ARMOR
        if howManyVisibleTreasures(TreasureKind::ARMOR)==0
          puede=true
        end
      end
      return puede
    end
    def howManyVisibleTreasures (tKind)
      suma = 0
      @visibleTreasures.each do |t|
        if t.type == t
          suma += 1
        end
      end
      return suma
    end
    def dieIfNoTreasures
      if @visibleTreasures.empty? && @hiddenTreasures.empty?
        @dead = true
      end
    end
    def combat (m)
      myLevel = getCombatLevel
      monsterLevel = getOponentLevel(m)
      if myLevel > monsterLevel
        applyPrize(m)
        if @level >= 10
          combatResult = CombatResult::WINGAME
        else
          combatResult = CombatResult::WIN
        end
      else
        applyBadConsequence(m)
        if shouldConvert
          combatResult = CombatResult::LOSEANDCONVERT
        else
          combatResult = CombatResult::LOSE
        end
      end
    end
    def makeTreasureVisible (t)
      canI = canMakeTreasureVisible(t)
      if canI
        @visibleTreasures << t
        @hiddenTreasures.delete(t)
      end
    end
    def discardVisibleTreasure (t)
      @visibleTreasures.delete(t)
      CardDealer.instance.giveTreasureBack(t)
      if @pendingBadConsequence!=nil && !@pendingBadConsequence.isEmpty
        @pendingBadConsequence.substractVisibleTreasure(t)
      end
      dieIfNoTreasures
    end
    def discardHiddenTreasure (t)
      @hiddenTreasures.delete(t)
      CardDealer.instance.giveTreasureBack(t)
      if @pendingBadConsequence!=nil && !@pendingBadConsequence.isEmpty
        @pendingBadConsequence.substractHiddenTreasure(t)
      end

      dieIfNoTreasures
    end
    def validState
      @pendingBadConsequence.isEmpty && (@hiddenTreasures.length <= 4)
    end
    def initTreasures
      dealer=CardDealer.instance
      dice=Dice.instance
      bringToLife
      treasure=dealer.nextTreasure
      @hiddenTreasures << treasure
      num = dice.nextNumber
      if num==6
        treasure =dealer.nextTreasure
        @hiddenTreasures << treasure
        treasure =dealer.nextTreasure
        @hiddenTreasures << treasure
      end
      if num>1 && num<6
        treasure =dealer.nextTreasure
        @hiddenTreasures << treasure
      end
    end
    def to_s
      "#{@name}"
    end
    def stealTreasure
      canI=canISteal
      treasure=nil
      if canI==true
        canYou=canYouGiveMeATreasure
        if canYou==true
          treasure = @enemy.giveMeATreasure
          @hiddenTreasures << treasure
          haveStolen
        end
      end
      return treasure
    end
     def giveMeATreasure
      num=@hiddenTreasures.size
      n=rand(num)
      return @hiddenTreasures[n]
    end
    def canYouGiveMeATreasure
      tiene=false
      if @hiddenTreasures.size >0
        tiene=true
      end
      return tiene
    end
    def haveStolen
       @canISteal=false
    end
    def discardAllTreasures
      auxVisible = Array.new(@visibleTreasures)
      auxHidden = Array.new(@hiddenTreasures)

      auxVisible.each do |t|
        discardVisibleTreasure(t)
      end

      auxHidden.each do |t|
        discardHiddenTreasure(t)
      end
    end
    protected :getCombatLevel, :getOponentLevel, :shouldConvert
    private :bringToLife, :incrementLevels, :decrementLevels, :setPendingBadConsequence, :applyPrize, :applyBadConsequence, :canMakeTreasureVisible, :howManyVisibleTreasures, :dieIfNoTreasures
  end
end
